lib/semantics/resolve-labels.cc

```diff
@@ -892,18 +892,20 @@ void CheckLabelDoConstraints(const SourceStmtList &dos,
       errorHandler.Say(position,
           parser::MessageFormattedText{
               "label '%u' is not in scope"_en_US, SayLabel(label)});
-    } else if (!doTarget.labeledStmtClassificationSet.test(
-                   TargetStatementEnum::Do) &&
-        !doTarget.labeledStmtClassificationSet.test(
-            TargetStatementEnum::CompatibleDo)) {
-      errorHandler.Say(doTarget.parserCharBlock,
-          parser::MessageFormattedText{
-              "'%u' invalid DO terminal statement"_err_en_US, SayLabel(label)});
     } else if (!doTarget.labeledStmtClassificationSet.test(
                    TargetStatementEnum::Do)) {
-      errorHandler.Say(doTarget.parserCharBlock,
-          parser::MessageFormattedText{
-              "'%u' invalid DO terminal statement"_en_US, SayLabel(label)});
+      if (!doTarget.labeledStmtClassificationSet.test(
+              TargetStatementEnum::CompatibleDo)) {
+        errorHandler.Say(doTarget.parserCharBlock,
+            parser::MessageFormattedText{
+                "'%u' DO statements must terminate with END DO or CONTINUE"_err_en_US,
```

Comment:
The thing being terminated is a DO loop, not a DO statement.

Reply:
Thanks.  I'll fix that.

Reply:
This is confusing.  Specifically, the "must" part of the new messages were never the intention.

The original code is testing two things which are being spliced together here.  In the former case, the test is if the DO loop ends with an invalid statement altogether, and reports an _error_ in that case.  In the latter case, the test is for Fortran 2018 compliance that the DO loop ends with a CONTINUE or END DO.  This is a _warning_ only, because of the project-level goal of allowing F18 to compile legal Fortran programs as defined by previous standards of the language.

Reply:
I didn't change the analysis in the compiler that distinguishes between the error and warning cases.  Also, there is the third case of not producing a message at all.  This change is simply to improve the message that comes out in the warning and error cases by adding additional information about how to make the message go away.  In the current state of the compiler, only the warning will come out.  For that warning case, the latest version of the code produces the message "A DO loop should terminate with END DO or CONTINUE".  Is this a good change to make at this time?